<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="application/xhtml+xml; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.6.8" />
<title>HPC GridRunner</title>
<style type="text/css">
/* Shared CSS for AsciiDoc xhtml11 and html5 backends */

/* Default font. */
body {
  font-family: Georgia,serif;
}

/* Title font. */
h1, h2, h3, h4, h5, h6,
div.title, caption.title,
thead, p.table.header,
#toctitle,
#author, #revnumber, #revdate, #revremark,
#footer {
  font-family: Arial,Helvetica,sans-serif;
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}
h5 {
  font-size: 1.0em;
}

div.sectionbody {
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}
ul > li     { color: #aaa; }
ul > li > * { color: black; }

.monospaced, code, pre {
  font-family: "Courier New", Courier, monospace;
  font-size: inherit;
  color: navy;
  padding: 0;
  margin: 0;
}


#author {
  color: #527bbd;
  font-weight: bold;
  font-size: 1.1em;
}
#email {
}
#revnumber, #revdate, #revremark {
}

#footer {
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

#preamble {
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.0em;
  margin-bottom: 2.0em;
  margin-right: 10%;
  color: #606060;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid #dddddd;
  border-left: 4px solid #f0f0f0;
  padding: 0.5em;
}

div.listingblock > div.content {
  border: 1px solid #dddddd;
  border-left: 5px solid #f0f0f0;
  background: #f8f8f8;
  padding: 0.5em;
}

div.quoteblock, div.verseblock {
  padding-left: 1.0em;
  margin-left: 1.0em;
  margin-right: 10%;
  border-left: 5px solid #f0f0f0;
  color: #888;
}

div.quoteblock > div.attribution {
  padding-top: 0.5em;
  text-align: right;
}

div.verseblock > pre.content {
  font-family: inherit;
  font-size: inherit;
}
div.verseblock > div.attribution {
  padding-top: 0.75em;
  text-align: left;
}
/* DEPRECATED: Pre version 8.2.7 verse style literal block. */
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 3px solid #dddddd;
}

div.exampleblock > div.content {
  border-left: 3px solid #dddddd;
  padding-left: 0.5em;
}

div.imageblock div.content { padding-left: 0; }
span.image img { border-style: none; }
a.image:visited { color: white; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
  color: navy;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
ol.arabic {
  list-style-type: decimal;
}
ol.loweralpha {
  list-style-type: lower-alpha;
}
ol.upperalpha {
  list-style-type: upper-alpha;
}
ol.lowerroman {
  list-style-type: lower-roman;
}
ol.upperroman {
  list-style-type: upper-roman;
}

div.compact ul, div.compact ol,
div.compact p, div.compact p,
div.compact div, div.compact div {
  margin-top: 0.1em;
  margin-bottom: 0.1em;
}

tfoot {
  font-weight: bold;
}
td > div.verse {
  white-space: pre;
}

div.hdlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hdlist tr {
  padding-bottom: 15px;
}
dt.hdlist1.strong, td.hdlist1.strong {
  font-weight: bold;
}
td.hdlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
  color: navy;
}
td.hdlist2 {
  vertical-align: top;
}
div.hdlist.compact tr {
  margin: 0;
  padding-bottom: 0;
}

.comment {
  background: yellow;
}

.footnote, .footnoteref {
  font-size: 0.8em;
}

span.footnote, span.footnoteref {
  vertical-align: super;
}

#footnotes {
  margin: 20px 0 20px 0;
  padding: 7px 0 0 0;
}

#footnotes div.footnote {
  margin: 0 0 5px 0;
}

#footnotes hr {
  border: none;
  border-top: 1px solid silver;
  height: 1px;
  text-align: left;
  margin-left: 0;
  width: 20%;
  min-width: 100px;
}

div.colist td {
  padding-right: 0.5em;
  padding-bottom: 0.3em;
  vertical-align: top;
}
div.colist td img {
  margin-top: 0.3em;
}

@media print {
  #footer-badges { display: none; }
}

#toc {
  margin-bottom: 2.5em;
}

#toctitle {
  color: #527bbd;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel0, div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}

span.aqua { color: aqua; }
span.black { color: black; }
span.blue { color: blue; }
span.fuchsia { color: fuchsia; }
span.gray { color: gray; }
span.green { color: green; }
span.lime { color: lime; }
span.maroon { color: maroon; }
span.navy { color: navy; }
span.olive { color: olive; }
span.purple { color: purple; }
span.red { color: red; }
span.silver { color: silver; }
span.teal { color: teal; }
span.white { color: white; }
span.yellow { color: yellow; }

span.aqua-background { background: aqua; }
span.black-background { background: black; }
span.blue-background { background: blue; }
span.fuchsia-background { background: fuchsia; }
span.gray-background { background: gray; }
span.green-background { background: green; }
span.lime-background { background: lime; }
span.maroon-background { background: maroon; }
span.navy-background { background: navy; }
span.olive-background { background: olive; }
span.purple-background { background: purple; }
span.red-background { background: red; }
span.silver-background { background: silver; }
span.teal-background { background: teal; }
span.white-background { background: white; }
span.yellow-background { background: yellow; }

span.big { font-size: 2em; }
span.small { font-size: 0.6em; }

span.underline { text-decoration: underline; }
span.overline { text-decoration: overline; }
span.line-through { text-decoration: line-through; }

div.unbreakable { page-break-inside: avoid; }


/*
 * xhtml11 specific
 *
 * */

div.tableblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
div.tableblock > table {
  border: 3px solid #527bbd;
}
thead, p.table.header {
  font-weight: bold;
  color: #527bbd;
}
p.table {
  margin-top: 0;
}
/* Because the table frame attribute is overriden by CSS in most browsers. */
div.tableblock > table[frame="void"] {
  border-style: none;
}
div.tableblock > table[frame="hsides"] {
  border-left-style: none;
  border-right-style: none;
}
div.tableblock > table[frame="vsides"] {
  border-top-style: none;
  border-bottom-style: none;
}


/*
 * html5 specific
 *
 * */

table.tableblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
thead, p.tableblock.header {
  font-weight: bold;
  color: #527bbd;
}
p.tableblock {
  margin-top: 0;
}
table.tableblock {
  border-width: 3px;
  border-spacing: 0px;
  border-style: solid;
  border-color: #527bbd;
  border-collapse: collapse;
}
th.tableblock, td.tableblock {
  border-width: 1px;
  padding: 4px;
  border-style: solid;
  border-color: #527bbd;
}

table.tableblock.frame-topbot {
  border-left-style: hidden;
  border-right-style: hidden;
}
table.tableblock.frame-sides {
  border-top-style: hidden;
  border-bottom-style: hidden;
}
table.tableblock.frame-none {
  border-style: hidden;
}

th.tableblock.halign-left, td.tableblock.halign-left {
  text-align: left;
}
th.tableblock.halign-center, td.tableblock.halign-center {
  text-align: center;
}
th.tableblock.halign-right, td.tableblock.halign-right {
  text-align: right;
}

th.tableblock.valign-top, td.tableblock.valign-top {
  vertical-align: top;
}
th.tableblock.valign-middle, td.tableblock.valign-middle {
  vertical-align: middle;
}
th.tableblock.valign-bottom, td.tableblock.valign-bottom {
  vertical-align: bottom;
}


/*
 * manpage specific
 *
 * */

body.manpage h1 {
  padding-top: 0.5em;
  padding-bottom: 0.5em;
  border-top: 2px solid silver;
  border-bottom: 2px solid silver;
}
body.manpage h2 {
  border-style: none;
}
body.manpage div.sectionbody {
  margin-left: 3em;
}

@media print {
  body.manpage div#toc { display: none; }
}


</style>
<script type="text/javascript">
/*<![CDATA[*/
var asciidoc = {  // Namespace.

/////////////////////////////////////////////////////////////////////
// Table Of Contents generator
/////////////////////////////////////////////////////////////////////

/* Author: Mihai Bazon, September 2002
 * http://students.infoiasi.ro/~mishoo
 *
 * Table Of Content generator
 * Version: 0.4
 *
 * Feel free to use this script under the terms of the GNU General Public
 * License, as long as you do not remove or alter this notice.
 */

 /* modified by Troy D. Hanson, September 2006. License: GPL */
 /* modified by Stuart Rackham, 2006, 2009. License: GPL */

// toclevels = 1..4.
toc: function (toclevels) {

  function getText(el) {
    var text = "";
    for (var i = el.firstChild; i != null; i = i.nextSibling) {
      if (i.nodeType == 3 /* Node.TEXT_NODE */) // IE doesn't speak constants.
        text += i.data;
      else if (i.firstChild != null)
        text += getText(i);
    }
    return text;
  }

  function TocEntry(el, text, toclevel) {
    this.element = el;
    this.text = text;
    this.toclevel = toclevel;
  }

  function tocEntries(el, toclevels) {
    var result = new Array;
    var re = new RegExp('[hH]([1-'+(toclevels+1)+'])');
    // Function that scans the DOM tree for header elements (the DOM2
    // nodeIterator API would be a better technique but not supported by all
    // browsers).
    var iterate = function (el) {
      for (var i = el.firstChild; i != null; i = i.nextSibling) {
        if (i.nodeType == 1 /* Node.ELEMENT_NODE */) {
          var mo = re.exec(i.tagName);
          if (mo && (i.getAttribute("class") || i.getAttribute("className")) != "float") {
            result[result.length] = new TocEntry(i, getText(i), mo[1]-1);
          }
          iterate(i);
        }
      }
    }
    iterate(el);
    return result;
  }

  var toc = document.getElementById("toc");
  if (!toc) {
    return;
  }

  // Delete existing TOC entries in case we're reloading the TOC.
  var tocEntriesToRemove = [];
  var i;
  for (i = 0; i < toc.childNodes.length; i++) {
    var entry = toc.childNodes[i];
    if (entry.nodeName.toLowerCase() == 'div'
     && entry.getAttribute("class")
     && entry.getAttribute("class").match(/^toclevel/))
      tocEntriesToRemove.push(entry);
  }
  for (i = 0; i < tocEntriesToRemove.length; i++) {
    toc.removeChild(tocEntriesToRemove[i]);
  }

  // Rebuild TOC entries.
  var entries = tocEntries(document.getElementById("content"), toclevels);
  for (var i = 0; i < entries.length; ++i) {
    var entry = entries[i];
    if (entry.element.id == "")
      entry.element.id = "_toc_" + i;
    var a = document.createElement("a");
    a.href = "#" + entry.element.id;
    a.appendChild(document.createTextNode(entry.text));
    var div = document.createElement("div");
    div.appendChild(a);
    div.className = "toclevel" + entry.toclevel;
    toc.appendChild(div);
  }
  if (entries.length == 0)
    toc.parentNode.removeChild(toc);
},


/////////////////////////////////////////////////////////////////////
// Footnotes generator
/////////////////////////////////////////////////////////////////////

/* Based on footnote generation code from:
 * http://www.brandspankingnew.net/archive/2005/07/format_footnote.html
 */

footnotes: function () {
  // Delete existing footnote entries in case we're reloading the footnodes.
  var i;
  var noteholder = document.getElementById("footnotes");
  if (!noteholder) {
    return;
  }
  var entriesToRemove = [];
  for (i = 0; i < noteholder.childNodes.length; i++) {
    var entry = noteholder.childNodes[i];
    if (entry.nodeName.toLowerCase() == 'div' && entry.getAttribute("class") == "footnote")
      entriesToRemove.push(entry);
  }
  for (i = 0; i < entriesToRemove.length; i++) {
    noteholder.removeChild(entriesToRemove[i]);
  }

  // Rebuild footnote entries.
  var cont = document.getElementById("content");
  var spans = cont.getElementsByTagName("span");
  var refs = {};
  var n = 0;
  for (i=0; i<spans.length; i++) {
    if (spans[i].className == "footnote") {
      n++;
      var note = spans[i].getAttribute("data-note");
      if (!note) {
        // Use [\s\S] in place of . so multi-line matches work.
        // Because JavaScript has no s (dotall) regex flag.
        note = spans[i].innerHTML.match(/\s*\[([\s\S]*)]\s*/)[1];
        spans[i].innerHTML =
          "[<a id='_footnoteref_" + n + "' href='#_footnote_" + n +
          "' title='View footnote' class='footnote'>" + n + "</a>]";
        spans[i].setAttribute("data-note", note);
      }
      noteholder.innerHTML +=
        "<div class='footnote' id='_footnote_" + n + "'>" +
        "<a href='#_footnoteref_" + n + "' title='Return to text'>" +
        n + "</a>. " + note + "</div>";
      var id =spans[i].getAttribute("id");
      if (id != null) refs["#"+id] = n;
    }
  }
  if (n == 0)
    noteholder.parentNode.removeChild(noteholder);
  else {
    // Process footnoterefs.
    for (i=0; i<spans.length; i++) {
      if (spans[i].className == "footnoteref") {
        var href = spans[i].getElementsByTagName("a")[0].getAttribute("href");
        href = href.match(/#.*/)[0];  // Because IE return full URL.
        n = refs[href];
        spans[i].innerHTML =
          "[<a href='#_footnote_" + n +
          "' title='View footnote' class='footnote'>" + n + "</a>]";
      }
    }
  }
},

install: function(toclevels) {
  var timerId;

  function reinstall() {
    asciidoc.footnotes();
    if (toclevels) {
      asciidoc.toc(toclevels);
    }
  }

  function reinstallAndRemoveTimer() {
    clearInterval(timerId);
    reinstall();
  }

  timerId = setInterval(reinstall, 500);
  if (document.addEventListener)
    document.addEventListener("DOMContentLoaded", reinstallAndRemoveTimer, false);
  else
    window.onload = reinstallAndRemoveTimer;
}

}
asciidoc.install();
/*]]>*/
</script>
</head>
<body class="article">
<div id="header">
<h1>HPC GridRunner</h1>
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph"><p>HPC GridRunner is a simple command-line interface to high throughput computing using a variety of different grid computing platforms, including LSF, SGE, SLURM, and PBS.  You simply create a file containing the many unix commands that you want to have executed, where each command had no dependencies on any other command, and HPC GridRunner will batch and launch the commands for execution on your compute farm.  The execution status (success, failure) is tracked for each command, so that if any individual command should fail, it will be flagged and identified to the user for further investigation.</p></div>
<div class="paragraph"><p>HPC GridRunner includes generic support for Bioinformatics applications (see below), where the user would like to run some bioinformatics application on each sequence included in a large multi-fasta file, such as a blast search or a hmmer Pfam search.  HPC GridRunner facilitates this by batching sequences to your liking and executing each blast or pfam search in parallel.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="_installing_hpc_gridrunner">Installing HPC GridRunner</h2>
<div class="sectionbody">
<div class="paragraph"><p>Download HPC GridRunner <a href="https://github.com/HpcGridRunner/HpcGridRunner.github.io/releases">here</a>.</p></div>
<div class="paragraph"><p>The system runs as a set of Perl scripts, so there&#8217;s no need to compile anything.</p></div>
<div class="paragraph"><p>Additional support is included to leverage <a href="http://parafly.sourceforge.net/">ParaFly</a> to attempt to re-execute any grid-failed commands locally (from within the HpcGridRunner process instead of on the computing grid node where jobs were dispatched).  This is useful in cases where some small subset of grid jobs might time-out or need more RAM than allocated on the grid node. If you want to leverage this option, be sure to also install <a href="http://parafly.sourceforge.net/">ParaFly</a>.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="_configuration">Configuration</h2>
<div class="sectionbody">
<div class="paragraph"><p>HPC GridRunner uses a configuration file system to determine how to interact with your grid computing platform.  Example configuration files can be found in the included <em>hpc_conf/</em> directory.  Each conf file includes just a handful of attributes that enable the system to work, including:</p></div>
<div class="ulist"><ul>
<li>
<p>
grid : the type of grid computing platform being used. (LSF, SGE, PBS, or SLURM)  See an example conf for your system.
</p>
</li>
<li>
<p>
cmd : the basic structure of a job submission (ie. using qsub or bsub), indicating the queue to submit to, memory cap, walltime, etc., basically everything that would show up in your command line to submit the job to the grid sans the actual command that would be executed on the node.
</p>
</li>
<li>
<p>
max_nodes : the maximum number of nodes to try to leverage simultaneously on your grid platform. You might set this to a low number if you dont want to oversubscribe to resources, particularly if you have high-RAM requirements.  This is up to you. You could leave it as a high number and just let the grid platform itself determine how many jobs should RUN vs. PEND.  The total (RUN+PEND) will equal this max_nodes parameter.
</p>
</li>
<li>
<p>
cmds_per_node : your unix commands to be run by HPC GridRunner will be batched according to this number of commands per job submission.  So, if you set this to 100, then all 100 commands will be submitted in a single bsub or qsub to the grid. Note, although they are batched, each command will be separately monitored by HPC GridRunner for success/failure status.
</p>
</li>
</ul></div>
<div class="paragraph"><p>The contents of an example configuration file <em>hpc_conf/Broad_LSF.regev.100.conf</em> is shown below:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>#-------------------------------------------------------------------------------------------
# grid type:
grid=LSF</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code># template for a grid submission
cmd=bsub -q regevlab -R "rusage[mem=10]"
# note -e error.file -o out.file are set internally, so dont set them in the above cmd.</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code># uses the LSF feature to pre-exec and check that the file system is mounted before executing.
# this helps when you have some misbehaving grid nodes that lost certain file mounts.
mount_test=T</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>##########################################################################################
# settings below configure the Trinity job submission system, not tied to the grid itself.
##########################################################################################</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code># number of grid submissions to be maintained at steady state by the Trinity submission system
max_nodes=500</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code># number of commands that are batched into a single grid submission job.
cmds_per_node=100</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>#--------------------------------------------------------------------------------------------</code></pre>
</div></div>
</div>
</div>
<div class="sect1">
<h2 id="_running_hpc_gridrunner">Running HPC GridRunner</h2>
<div class="sectionbody">
<div class="paragraph"><p>The simplest way to run HPC GridRunner is to have a file containing a list of unix commands to be executed, and to give that file as a parameter to the <em>hpc_cmds_GridRunner.pl</em> script.  Usage is shown below:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>./hpc_cmds_GridRunner.pl</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>################################################################
# Required:
#
#  -c &lt;string&gt;        file containing list of commands
#  --grid_conf|G &lt;string&gt;   grid config file
#
# Optional:
#
#  --parafly          if any grid commands fail on the grid, try rerunning
#                     them locally using ParaFly (second chance to succeed).
#                     This requires that ParaFly be installed and in your PATH
#                     Get ParaFly here: http://parafly.sourceforge.net/
#
####################################################################</code></pre>
</div></div>
<div class="paragraph"><p>For example, imagine a command file called <em>myCommands.txt</em> that contains the following:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>echo hello1
echo hello2
echo hello3
echoXX hello4</code></pre>
</div></div>
<div class="paragraph"><p>You would execute this file of commands on your computing grid like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>hpc_cmds_GridRunner.pl -c myCommands.txt -G /path/to/grid.conf</code></pre>
</div></div>
<div class="paragraph"><p>Your grid.conf file could be set up to have <em>cmds_per_node=1</em>, such that each command line in that text file will be launched on a different compute node.</p></div>
<div class="paragraph"><p>The results would look like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>bhaas@tin $ ~/GITHUB/HpcGridRunner/hpc_cmds_GridRunner.pl  -c ./myCommands.txt -G ~/GITHUB/HpcGridRunner/hpc_conf/BroadInst_LSF.hour.1.conf
SERVER: tin.broadinstitute.org, PID: 24527
 $VAR1 = {
       'mount_test' =&gt; 'T',
       'grid' =&gt; 'LSF',
       'cmd' =&gt; 'bsub -q hour -W 4:0 -R "rusage[mem=10]"',
       'max_nodes' =&gt; '500',
       'cmds_per_node' =&gt; '1'
     };
CMDS: 4 / 4  [4/500 nodes in use]
* All cmds submitted to grid.  Now waiting for them to finish.</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>Failures encountered:
num_success: 3  num_fail: 1     num_unknown: 0
Finished.</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>1 commands failed during grid computing.
-failed commands written to: ./myCommands.txt.hpc-cache_success.__failures</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>Error, not all commands could complete successfully... cannot continue.</code></pre>
</div></div>
<div class="paragraph"><p>I see that 1 of the jobs failed, and check the contents of the corresponding file mentioned directly above:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>bhaas@tin $ cat ./myCommands.txt.hpc-cache_success.__failures
    echoXX hello4</code></pre>
</div></div>
<div class="paragraph"><p>and it&#8217;s clear that my command <em>echoXX hello4</em> failed&#8230;  and of course, I made that command to fail on purpose as there is no echoXX command, and so this serves to demonstrate what to expect when you have some subset of jobs that fail in execution.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="_another_example_more_complex_unix_commands">Another example, more complex unix commands</h2>
<div class="sectionbody">
<div class="paragraph"><p>I have another file containing a set of unix commands, and they&#8217;re slightly more complicated in that they include multiple piped-together commands and we want to capture the output.</p></div>
<div class="paragraph"><p>My <em>commands_ex2.txt</em> file has the following commands:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>echo a | wc -c &gt; a.len
echo ab | wc -c &gt; ab.len
echo abc | wc -c &gt; abc.len
echo hello world | wc &gt; h.whatever</code></pre>
</div></div>
<div class="paragraph"><p>Running the system:</p></div>
<div class="literalblock">
<div class="content">
<pre><code> bhaas@tin $ ~/GITHUB/HpcGridRunner/hpc_cmds_GridRunner.pl  -c ./commands_ex2.txt -G ~/GITHUB/HpcGridRunner/hpc_conf/BroadInst_LSF.hour.1.conf
 SERVER: tin.broadinstitute.org, PID: 19267
 $VAR1 = {
        'mount_test' =&gt; 'T',
        'grid' =&gt; 'LSF',
        'cmd' =&gt; 'bsub -q hour -W 4:0 -R "rusage[mem=10]"',
        'max_nodes' =&gt; '500',
        'cmds_per_node' =&gt; '1'
      };
CMDS: 4 / 4  [4/500 nodes in use]
* All cmds submitted to grid.  Now waiting for them to finish.
CMDS: 4 / 4  [0/500 nodes in use]
* All nodes completed.  Now auditing job completion status values
All commands completed successfully.
Finished.</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>All commands completed successfully on the computing grid.</code></pre>
</div></div>
<div class="paragraph"><p>This time, each command completed successfully.  If I look at the contents of my working directory, I see that files were generated according to the commands above:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>-rw-rw-rw- 1 bhaas broad    2 Jan 26 10:37 ab.len
-rw-rw-rw- 1 bhaas broad    2 Jan 26 10:37 abc.len
-rw-rw-rw- 1 bhaas broad   24 Jan 26 10:37 h.whatever
-rw-rw-rw- 1 bhaas broad    2 Jan 26 10:37 a.len</code></pre>
</div></div>
<div class="paragraph"><p>And I can verify their contents:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>bhaas@tin $ cat *len
2
3
4</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>bhaas@tin $ cat h.whatever
   1       2      12</code></pre>
</div></div>
<div class="paragraph"><p>So, the system is very flexibile and is designed to execute exactly what you indicate as a unix command in your file of commands.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="_using_hpc_gridrunner_for_bioinformatics_applications">Using HPC GridRunner for Bioinformatics Applications</h2>
<div class="sectionbody">
<div class="paragraph"><p>Included with HPC GridRunner is a utility to perform similar generic processing of sequences provided in a multi-fasta file. This is enabled by the script:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>./BioIfx/hpc_FASTA_GridRunner.pl</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>#######################################################################################################################
#
# Required:
#
#   --query_fasta|Q &lt;string&gt;       query multiFastaFile (full or relative path)
#
#   --cmd_template|T &lt;string&gt;      program command line template:   eg. "/path/to/prog [opts] __QUERY_FILE__ [other opts]"
#
#  --grid_conf|G &lt;string&gt;          grid config file (see hpc_conf/ for examples)
#
#   --seqs_per_bin|N &lt;int&gt;         number of sequences per partition.
#
#   --out_dir|O &lt;string&gt;           output directory
#
# Optional:
#
#   --prep_only|X                  partion data and create cmds list, but don't launch on the grid.
#
#   --parafly                      use parafly to re-exec previously failed grid commands
#
########################################################################################################################</code></pre>
</div></div>
<div class="paragraph"><p>The <em>--cmd_template</em> parameter enables you to define a command that would be executed on each set of sequences in your input query fasta file. The sequences are batched up into smaller multi-fasta files, each containing <em>--seqs_per_bin</em> number of sequences, the command described in the <em>--cmd_template</em> is executed on that partition of sequences, and the output is stored in the <em>--out_dir</em> specified output directory.</p></div>
<div class="sect2">
<h3 id="_example_running_blast">Example running BLAST</h3>
<div class="admonitionblock">
<table><tr>
<td class="icon">
<div class="title">Note</div>
</td>
<td class="content">Be sure to already have <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">BLAST+</a> installed.</td>
</tr></table>
</div>
<div class="paragraph"><p>An example of using this to run a blastp search of a set of protein sequences might look like so.</p></div>
<div class="literalblock">
<div class="content">
<pre><code>./BioIfx/hpc_FASTA_GridRunner.pl \
--cmd_template "blastp -query __QUERY_FILE__ -db /seq/RNASEQ/DBs/SWISSPROT/current/uniprot_sprot.fasta  -max_target_seqs 1 -outfmt 6 -evalue 1e-5" \
--query_fasta test.pep \
-G ../hpc_conf/BroadInst_LSF.test.conf \
-N 10 -O test_blastp_search</code></pre>
</div></div>
<div class="paragraph"><p>which would report to the terminal:</p></div>
<div class="literalblock">
<div class="content">
<pre><code> Sequences to search: test_blastp_search/grp_0001/1.fa test_blastp_search/grp_0001/2.fa test_blastp_search/grp_0001/3.fa test_blastp_search/grp_0001/4.fa test_blastp_search/grp_0001/5.fa test_blastp_search/grp_0001/6.fa test_blastp_search/grp_0001/7.fa test_blastp_search/grp_0001/8.fa test_blastp_search/grp_0001/9.fa test_blastp_search/grp_0001/10.fa
There are 10 jobs to run.
$VAR1 = {
        'mount_test' =&gt; 'T',
        'grid' =&gt; 'LSF',
        'cmd' =&gt; 'bsub -q regevlab -R "rusage[mem=2]" ',
        'max_nodes' =&gt; '500',
        'cmds_per_node' =&gt; '1'
      };
CMDS: 10 / 10  [10/500 nodes in use]
* All cmds submitted to grid.  Now waiting for them to finish.</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>CMDS: 10 / 10  [0/500 nodes in use]
* All nodes completed.  Now auditing job completion status values
All commands completed successfully.
Finished.</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>All commands completed successfully on the computing grid.
SUCCESS:  all commands completed succesfully. :)</code></pre>
</div></div>
<div class="paragraph"><p>All the output was set to be placed in a <em>test_blastp_search/</em> directory, according to my command above.  This directory was automatically created and I can examine the contents like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>bhaas@venustum $ find test_blastp_search
test_blastp_search
test_blastp_search/grp_0001
test_blastp_search/grp_0001/10.fa.ERR
test_blastp_search/grp_0001/2.fa
test_blastp_search/grp_0001/1.fa.ERR
test_blastp_search/grp_0001/5.fa.ERR
test_blastp_search/grp_0001/3.fa
test_blastp_search/grp_0001/8.fa.ERR
test_blastp_search/grp_0001/6.fa.ERR
test_blastp_search/grp_0001/7.fa.ERR
test_blastp_search/grp_0001/10.fa.OUT
test_blastp_search/grp_0001/3.fa.ERR
test_blastp_search/grp_0001/4.fa.ERR
test_blastp_search/grp_0001/6.fa
test_blastp_search/grp_0001/9.fa.ERR
test_blastp_search/grp_0001/7.fa.OUT
test_blastp_search/grp_0001/6.fa.OUT
test_blastp_search/grp_0001/4.fa.OUT
test_blastp_search/grp_0001/5.fa
test_blastp_search/grp_0001/8.fa.OUT
...</code></pre>
</div></div>
<div class="paragraph"><p>My original <em>test.pep</em> multi-fasta file was partitioned into a number of .fa files shown above, with each .fa file containing 10 sequences (according to my <em>-N 10</em> parameter setting).  Blastp was executed on each .fa file, with the output captured in each .fa.OUT file, and any error messages (anything blast writes to the stderr stream) captured as .fa.ERR.</p></div>
<div class="paragraph"><p>To capture all my blast results and put them in a single file, I would execute the following:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>find test_blastp_search -name "*.fa.OUT" -exec cat {} \; &gt; all.blast.out</code></pre>
</div></div>
<div class="paragraph"><p>and now I have a single output containing all the blast results for further study.</p></div>
</div>
<div class="sect2">
<h3 id="_example_running_pfam_searches">Example running Pfam searches</h3>
<div class="admonitionblock">
<table><tr>
<td class="icon">
<div class="title">Note</div>
</td>
<td class="content">Be sure to have <a href="http://hmmer.janelia.org/">hmmer3</a> and <a href="ftp://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz">Pfam</a> databases installed.</td>
</tr></table>
</div>
<div class="paragraph"><p>I can run a Pfam search like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>./BioIfx/hpc_FASTA_GridRunner.pl \
--cmd_template "hmmscan --cpu 8 --domtblout __QUERY_FILE__.domtblout /seq/RNASEQ/DBs/PFAM/current/Pfam-A.hmm __QUERY_FILE__" \
--query_fasta test.pep \
-G ../hpc_conf/BroadInst_LSF.test.conf \
-N 10 -O test_pfam_search</code></pre>
</div></div>
<div class="paragraph"><p>which reports to the terminal:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>Sequences to search: test_pfam_search/grp_0001/1.fa test_pfam_search/grp_0001/2.fa test_pfam_search/grp_0001/3.fa test_pfam_search/grp_0001/4.fa test_pfam_search/grp_0001/5.fa test_pfam_search/grp_0001/6.fa test_pfam_search/grp_0001/7.fa test_pfam_search/grp_0001/8.fa test_pfam_search/grp_0001/9.fa test_pfam_search/grp_0001/10.fa
There are 10 jobs to run.
$VAR1 = {
        'mount_test' =&gt; 'T',
        'grid' =&gt; 'LSF',
        'cmd' =&gt; 'bsub -q regevlab -R "rusage[mem=2]" ',
        'max_nodes' =&gt; '500',
        'cmds_per_node' =&gt; '1'
      };
CMDS: 10 / 10  [10/500 nodes in use]
* All cmds submitted to grid.  Now waiting for them to finish.
CMDS: 10 / 10  [0/500 nodes in use]
* All nodes completed.  Now auditing job completion status values
All commands completed successfully.
Finished.</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>All commands completed successfully on the computing grid.
SUCCESS:  all commands completed succesfully. :)</code></pre>
</div></div>
<div class="paragraph"><p>I find a <em>test_pfam_search/</em> directory created and that has the following contents:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>bhaas@venustum $ find test_pfam_search
test_pfam_search
test_pfam_search/grp_0001
test_pfam_search/grp_0001/5.fa.domtblout
test_pfam_search/grp_0001/3.fa.domtblout
test_pfam_search/grp_0001/4.fa.domtblout
test_pfam_search/grp_0001/10.fa.ERR
test_pfam_search/grp_0001/2.fa
test_pfam_search/grp_0001/1.fa.ERR
test_pfam_search/grp_0001/6.fa.domtblout
test_pfam_search/grp_0001/5.fa.ERR
test_pfam_search/grp_0001/3.fa
test_pfam_search/grp_0001/8.fa.ERR
test_pfam_search/grp_0001/8.fa.domtblout
test_pfam_search/grp_0001/1.fa.domtblout
test_pfam_search/grp_0001/6.fa.ERR
test_pfam_search/grp_0001/7.fa.ERR
test_pfam_search/grp_0001/10.fa.OUT
test_pfam_search/grp_0001/2.fa.domtblout
test_pfam_search/grp_0001/9.fa.domtblout
...</code></pre>
</div></div>
<div class="paragraph"><p>In this case, the command template directed hmmscan to create a .domtblout file for each .fa sequence set containing the Pfam domain hits.  We can capture these outputs and put them in a single file like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>find test_pfam_search/ -name "*.fa.domtblout" -exec cat {} \; &gt; all_pfam_results.out</code></pre>
</div></div>
<div class="paragraph"><p>and now I have all the pfam domain hits in a single file for use in downstream analyses or applications.</p></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_questions_comments_etc">Questions, comments, etc?</h2>
<div class="sectionbody">
<div class="paragraph"><p>Please use the google group: <a href="https://groups.google.com/forum/#!forum/hpcgridrunner">https://groups.google.com/forum/#!forum/hpcgridrunner</a></p></div>
</div>
</div>
</div>
<div id="footnotes"><hr /></div>
<div id="footer">
<div id="footer-text">
Last updated 2015-01-26 11:17:24 EST
</div>
</div>
</body>
</html>
